File: src/pip_mirror/indexer.py
"""PEP 503 Simple Repository 索引生成，支持 PEP 658 metadata 和 PEP 691 JSON API."""

from __future__ import annotations

import json
import logging
from pathlib import Path

from .sqlite_store import DownloadStore

logger = logging.getLogger("pip-mirror")


_INDEX_HTML_TEMPLATE = """<!DOCTYPE html>
<html>
<head>
    <meta charset="UTF-8">
    <title>Simple Index</title>
</head>
<body>
{links}
</body>
</html>
"""

_PACKAGE_HTML_TEMPLATE = """<!DOCTYPE html>
<html>
<head>
    <meta charset="UTF-8">
    <title>Links for {package_name}</title>
</head>
<body>
<h1>Links for {package_name}</h1>
{links}
</body>
</html>
"""


def _generate_index_html(package_names: list[str]) -> str:
    """生成根 index.html（包列表）."""
    lines = []
    for name in sorted(package_names):
        lines.append(f'    <a href="{name}/">{name}</a><br/>')
    return _INDEX_HTML_TEMPLATE.format(links="\n".join(lines))


def _generate_index_json(package_names: list[str]) -> str:
    """生成根 index.json（PEP 691）."""
    data = {
        "meta": {"api-version": "1.0"},
        "projects": [{"name": name} for name in sorted(package_names)],
    }
    return json.dumps(data, indent=2, ensure_ascii=False)


def _generate_package_html(
    package_name: str,
    files: list[Path],
    hashes: dict[str, str],
    metadata_hashes: dict[str, str],
) -> str:
    """生成单个包的 index.html，带 PEP 658 data- 属性."""
    lines = []
    for f in sorted(files):
        attrs = f'href="{f.name}"'
        sha256 = hashes.get(f.name)
        if sha256:
            attrs += f' data-sha256="{sha256}"'
        meta_sha256 = metadata_hashes.get(f.name)
        if meta_sha256 and f.name.endswith(".whl"):
            attrs += f' data-core-metadata="sha256={meta_sha256}"'
            attrs += f' data-dist-info-metadata="sha256={meta_sha256}"'
        lines.append(f"    <a {attrs}>{f.name}</a><br/>")
    return _PACKAGE_HTML_TEMPLATE.format(
        package_name=package_name,
        links="\n".join(lines),
    )


def _generate_package_json(
    package_name: str,
    files: list[Path],
    hashes: dict[str, str],
    metadata_hashes: dict[str, str],
) -> str:
    """生成单个包的 index.json（PEP 691）."""
    file_entries = []
    for f in sorted(files):
        entry: dict = {
            "filename": f.name,
            "url": f.name,
            "hashes": {},
        }
        sha256 = hashes.get(f.name)
        if sha256:
            entry["hashes"]["sha256"] = sha256
        meta_sha256 = metadata_hashes.get(f.name)
        if meta_sha256 and f.name.endswith(".whl"):
            entry["dist-info-metadata"] = {"sha256": meta_sha256}
        file_entries.append(entry)

    data = {
        "meta": {"api-version": "1.0"},
        "name": package_name,
        "files": file_entries,
    }
    return json.dumps(data, indent=2, ensure_ascii=False)


def generate_index(repository_dir: Path) -> None:
    """生成 PEP 503 / PEP 691 规范的 simple index."""
    simple_dir = repository_dir / "simple"
    if not simple_dir.exists():
        logger.info("仓库目录为空，跳过索引生成")
        return

    logger.info("生成 PEP 503 / PEP 691 索引...")

    store = DownloadStore(repository_dir / ".store.db")
    hashes = store.get_all_hashes()
    metadata_hashes = store.get_all_metadata_hashes()

    package_names: list[str] = []

    for pkg_dir in sorted(simple_dir.iterdir()):
        if not pkg_dir.is_dir():
            continue

        pkg_name = pkg_dir.name
        package_names.append(pkg_name)

        files = [
            f for f in pkg_dir.iterdir()
            if f.is_file()
            and not f.name.startswith(".")
            and not f.name.endswith(".tmp")
            and not f.name.endswith(".metadata")
        ]

        html = _generate_package_html(pkg_name, files, hashes, metadata_hashes)
        (pkg_dir / "index.html").write_text(html, encoding="utf-8")

        json_content = _generate_package_json(pkg_name, files, hashes, metadata_hashes)
        (pkg_dir / "index.json").write_text(json_content, encoding="utf-8")

    root_html = _generate_index_html(package_names)
    (simple_dir / "index.html").write_text(root_html, encoding="utf-8")

    root_json = _generate_index_json(package_names)
    (simple_dir / "index.json").write_text(root_json, encoding="utf-8")

    logger.info(f"索引生成完成: {len(package_names)} 个包")

File: src/indexer.rs
use std::path::Path;

use dashmap::DashMap;
use tracing::info;

use crate::progress::{ProgressHandle, SyncEvent};
use crate::store::DownloadStore;

static INDEX_HTML_TEMPLATE: &str = r#"<!DOCTYPE html>
<html><head><meta charset="UTF-8"><title>Simple Index</title></head>
<body>{links}</body></html>"#;

static PACKAGE_HTML_TEMPLATE: &str = r#"<!DOCTYPE html>
<html><head><meta charset="UTF-8"><title>Links for {package_name}</title></head>
<body><h1>Links for {package_name}</h1>{links}</body></html>"#;

fn collect_files(path: &std::path::Path) -> Vec<String> {
    let mut files: Vec<String> = std::fs::read_dir(path)
        .into_iter()
        .flatten()
        .flatten()
        .filter(|f| f.path().is_file())
        .map(|f| f.file_name().to_string_lossy().to_string())
        .filter(|n| {
            !n.starts_with('.')
                && !n.ends_with(".tmp")
                && !n.ends_with(".metadata")
        })
        .collect();
    files.sort();
    files
}

struct IndexPkg<'a> {
    path: &'a std::path::Path,
    name: &'a str,
    hashes: &'a DashMap<String, String>,
    meta: &'a DashMap<String, String>,
    yanked: &'a DashMap<String, String>,
}

fn index_pkg(ctx: &IndexPkg<'_>) {
    let files = collect_files(ctx.path);
    let pkg_ctx = PkgCtx {
        package_name: ctx.name,
        files: &files,
        hashes: ctx.hashes,
        metadata_hashes: ctx.meta,
        yanked: ctx.yanked,
    };
    let _ = std::fs::write(
        ctx.path.join("index.html"),
        generate_package_html(&pkg_ctx),
    );
    let _ = std::fs::write(
        ctx.path.join("index.json"),
        generate_package_json(&pkg_ctx),
    );
}

type StoreMaps = (
    DashMap<String, String>,
    DashMap<String, String>,
    DashMap<String, String>,
);

fn load_store_maps(repository_dir: &Path) -> StoreMaps {
    let db_path = repository_dir.join(".store.db");
    if let Ok(store) = DownloadStore::open(&db_path) {
        return (
            store.get_all_hashes(),
            store.get_all_metadata_hashes(),
            store.get_all_yanked(),
        );
    }
    info!(".store.db 不存在或无法打开，使用空 hash/yanked 生成索引");
    (DashMap::new(), DashMap::new(), DashMap::new())
}

fn list_package_dirs(simple_dir: &Path) -> Vec<std::fs::DirEntry> {
    let Ok(entries) = std::fs::read_dir(simple_dir) else {
        return Vec::new();
    };
    entries.flatten().filter(|e| e.path().is_dir()).collect()
}

fn write_root_indexes(simple_dir: &Path, names: &[String]) {
    let _ = std::fs::write(
        simple_dir.join("index.html"),
        generate_index_html(names),
    );
    let _ = std::fs::write(
        simple_dir.join("index.json"),
        generate_index_json(names),
    );
}

fn emit_index_started(progress: &ProgressHandle, total: usize) {
    progress.emit(SyncEvent::PhaseStarted {
        phase: "index",
        total: Some(total as u64),
    });
}

fn emit_index_progress(
    progress: &ProgressHandle,
    current: usize,
    message: String,
) {
    progress.emit(SyncEvent::PhaseProgress {
        phase: "index",
        current: current as u64,
        message,
    });
}

fn emit_index_finished(progress: &ProgressHandle, count: usize) {
    progress.emit(SyncEvent::PhaseFinished {
        phase: "index",
        summary: format!("{} 个包", count),
    });
}

fn index_packages(
    entries: &[std::fs::DirEntry],
    hashes: &DashMap<String, String>,
    meta: &DashMap<String, String>,
    yanked: &DashMap<String, String>,
    progress: Option<&ProgressHandle>,
) -> Vec<String> {
    let mut names: Vec<String> = Vec::with_capacity(entries.len());
    for (idx, entry) in entries.iter().enumerate() {
        let path = entry.path();
        let name = path.file_name().unwrap().to_string_lossy().to_string();
        names.push(name.clone());
        index_pkg(&IndexPkg {
            path: &path,
            name: &name,
            hashes,
            meta,
            yanked,
        });
        if let Some(p) = progress {
            emit_index_progress(p, idx + 1, name);
        }
    }
    names
}

pub fn generate_index(repository_dir: &Path, progress: Option<ProgressHandle>) {
    let simple_dir = repository_dir.join("simple");
    if !simple_dir.exists() {
        info!("仓库目录为空，跳过索引生成");
        return;
    }
    info!("生成 PEP 503 / PEP 691 索引...");

    let (hashes, meta, yanked) = load_store_maps(repository_dir);
    let entries = list_package_dirs(&simple_dir);

    if let Some(ref p) = progress {
        emit_index_started(p, entries.len());
    }

    let mut names =
        index_packages(&entries, &hashes, &meta, &yanked, progress.as_ref());
    names.sort();
    write_root_indexes(&simple_dir, &names);

    if let Some(ref p) = progress {
        emit_index_finished(p, names.len());
    }

    info!("索引生成完成: {} 个包", names.len());
}

fn generate_index_html(package_names: &[String]) -> String {
    let links: Vec<String> = package_names
        .iter()
        .map(|name| format!(r#"    <a href="{name}/">{name}</a><br/>"#))
        .collect();
    INDEX_HTML_TEMPLATE.replace("{links}", &links.join("\n"))
}

fn generate_index_json(package_names: &[String]) -> String {
    let projects: Vec<serde_json::Value> = package_names
        .iter()
        .map(|n| serde_json::json!({"name": n}))
        .collect();
    serde_json::to_string_pretty(&serde_json::json!({
        "meta": {"api-version": "1.0"},
        "projects": projects,
    }))
    .unwrap()
}

struct PkgCtx<'a> {
    package_name: &'a str,
    files: &'a [String],
    hashes: &'a DashMap<String, String>,
    metadata_hashes: &'a DashMap<String, String>,
    yanked: &'a DashMap<String, String>,
}

fn build_link_attrs(
    f: &str,
    hashes: &DashMap<String, String>,
    meta: &DashMap<String, String>,
    yanked: &DashMap<String, String>,
) -> String {
    let mut attrs = format!(r#"href="{f}""#);
    if let Some(sha) = hashes.get(f) {
        attrs.push_str(&format!(r#" data-sha256="{}""#, sha.value()));
    }
    if f.ends_with(".whl")
        && let Some(m) = meta.get(f)
    {
        attrs.push_str(&format!(
            r#" data-core-metadata="sha256={}" data-dist-info-metadata="sha256={}""#,
            m.value(),
            m.value()
        ));
    }
    if let Some(y) = yanked.get(f) {
        attrs.push_str(&format!(r#" data-yanked="{}""#, y.value()));
    }
    attrs
}

fn generate_package_html(ctx: &PkgCtx<'_>) -> String {
    let links: Vec<String> = ctx
        .files
        .iter()
        .map(|f| {
            let attrs = build_link_attrs(
                f,
                ctx.hashes,
                ctx.metadata_hashes,
                ctx.yanked,
            );
            format!(r#"    <a {attrs}>{f}</a><br/>"#)
        })
        .collect();
    PACKAGE_HTML_TEMPLATE
        .replace("{package_name}", ctx.package_name)
        .replace("{links}", &links.join("\n"))
}

fn generate_package_json(ctx: &PkgCtx<'_>) -> String {
    let file_entries: Vec<serde_json::Value> = ctx
        .files
        .iter()
        .map(|f| {
            let mut entry =
                serde_json::json!({"filename": f, "url": f, "hashes": {}});
            if let Some(sha) = ctx.hashes.get(f) {
                entry["hashes"]["sha256"] =
                    serde_json::Value::String(sha.value().clone());
            }
            if f.ends_with(".whl")
                && let Some(meta) = ctx.metadata_hashes.get(f)
            {
                entry["dist-info-metadata"] =
                    serde_json::json!({"sha256": meta.value()});
            }
            if let Some(y) = ctx.yanked.get(f) {
                entry["yanked"] = serde_json::Value::String(y.value().clone());
            }
            entry
        })
        .collect();

    serde_json::to_string_pretty(&serde_json::json!({
        "meta": {"api-version": "1.0"},
        "name": ctx.package_name,
        "files": file_entries,
    }))
    .unwrap()
}
